doc/developer-guide/architecture/threads-and-events.en.rst

```diff
@@ -0,0 +1,135 @@
+.. Licensed to the Apache Software Foundation (ASF) under one
+   or more contributor license agreements.  See the NOTICE file
+   distributed with this work for additional information
+   regarding copyright ownership.  The ASF licenses this file
+   to you under the Apache License, Version 2.0 (the
+   "License"); you may not use this file except in compliance
+   with the License.  You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing,
+   software distributed under the License is distributed on an
+   "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+   KIND, either express or implied.  See the License for the
+   specific language governing permissions and limitations
+   under the License.
+
+.. include:: ../../common.defs
+
+.. default-domain:: cpp
+
+.. highlight:: cpp
+
+.. _threads-and-events:
+
+Threads and Event Processing
+****************************
+
+Interally |TS| is a cooperative multi-threaded environment. There are a fixed number of threads for core operations, determined at process start time. All core operations take place on one of these existing threads. Plugins may spawn additional threads but these are outside the scope of this document.
+
+Threads
+=======
+
+|TS| has a taxonomy of thread types. At the most basic are threads as the operating system / system libraries provide. On top of this |TS| has a thing layer that is the :class:`Thread` which contains a mutex and a thread identifier. The logic for starting the thread at the system level is embedded in this class. All threads started by |TS| have an instance of this class (or subclass). Plugins can directly start their own threads via system calls and those are not tracked. :class:`Thread` sets up thread local storage via :code:`pthread_setspecific`. Threads can be started via an explicit function provided to :method:`Thread::start` or by subclassing :class:`Thread` and overriding :method:`Thread::execute`.
+
+:class:`EThread` is a subclass of :class:`Thread` which provides support for |TS| core operations. It is this class that provides support for using :class:`Continuation` instances. Each :class:`EThread` instance executes a single continuation at thread start. If the thread is :enumerator:`ThreadType::DEDICATED` it exits after invoking the start continuation. A :enumerator:`ThreadType::REGULAR` thread will process its event queue until explicitly stopped after executing the start continuation.
```

Comment:
>> If the thread is :enumerator:`ThreadType::DEDICATED` it exits after invoking the start continuation.

The DEDICATED EThread is not exits since no one calls pthread_join().
It is just stops and keeps its stack like a zombie process until the ATS process exits.